WordPress/gutenberg
# [RNMobile] Enable child block settings to be hidden via parent block

## Description

More and more blocks are beginning to use [the Inner Block API](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/inner-blocks) to pull in child blocks to a parent block. As part of this, it's sometimes necessary to change or hide certain parts of a child block.

For the native apps, the settings control is included within each block. This is different to the web, where the block's settings are separated out into a sidebar. Due to this difference, the ability to hide the settings control for child blocks is likely to be a fairly common use case for parent blocks on the native side.

With this PR, a `__experimentalHideChildBlockControls` [block support](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-supports/), is introduced. If a parent block declares `__experimentalHideChildBlockControls` to be `true` with this PR applied, then the settings control for its child blocks won't be displayed. It is set to `false` by default.

The support was inspired by other experimental block supports that allow developers to change aspects of child blocks, for example: Following discussions in https://github.com/WordPress/gutenberg/issues/26313, `__experimentalExposeControlsToChildren` and `__experimentalShareWithChildBlocks` were introduced in https://github.com/WordPress/gutenberg/pull/33955 and https://github.com/WordPress/gutenberg/pull/34644.

## How has this been tested?

_Note: The Columns block is being used as a proof of concept here and for the purposes of testing. We are not planning to actually remove the settings control from inner Column blocks. The following steps can be repurposed for any block using inner blocks._

* Add `"__experimentalHideChildBlockControls": true` to the columns' block's `block.json` file. It should be added to [the block supports object here](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/columns/block.json#L18).
* With the change saved, add a new Columns block to the editor in the Android or iOS apps.
* Edit an individual Column block within Columns and confirm that the settings control is **_not_** available for the individual inner blocks.

We should also test other blocks using inner blocks, such as Social Icons and Buttons, to verify there are no regressions.

## Screenshots 

| Before  | After |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/141150714-67f4f0bc-ee1e-461e-bef3-d18a270a5af9.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/141150437-ed92518b-a4ea-4d38-b2a0-1fdce31320c0.png" width="100%"> |

## Types of changes

This PR introduces a non-breaking change which fixes an issue, with the following high-level overview of the code:

* A native-specific version of the `use-display-block-controls` component was created in https://github.com/WordPress/gutenberg/pull/36371/commits/d902f41db3c6635385528292bc6d8c26203ac837.
* In https://github.com/WordPress/gutenberg/pull/36371/commits/a23280cbc7323c9fb640e7cf5f5e3d7449566c16, a check for whether `__experimentalHideChildBlockControls` is true was added to the newly created file. This will now enable parent blocks to set the value to `true` if the settings control should be hidden from child blocks. [`hasBlockSupport`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-blocks/#hasblocksupport) is used to achieve this.

## Checklist:
